File: Frontend/src/app/viewModels/requests/RegenerateRouteVM.ts
import { toJSONString } from "../interfaces/toJSONString";
import { GenerateRouteVM } from "./GenerateRouteVM";

/**
 * View Model class for a re-generate Route request
 */
export class RegenerateRouteVM extends GenerateRouteVM implements toJSONString {

    /**
     * Construct's an instance of {@link RegenerateRouteVM}
     * 
     * @param name route's name
     * @param startTime route's start date
     * @param finishTime route's finishs date
     * @param visitedPlaces Specifies if the generated route should have palces already visited.
     * @param categories route's categories
     * @param deletedPlaces places that were deleted by the user.
     */
    constructor(
        public name: string,
        public startTime: Date,
        public finishTime: Date,
        public visitedPlaces: Boolean,
        public categories: Array<string>,
        public deletedPlaces?: Array<string>
    ) {
        super(name, startTime, finishTime, visitedPlaces, categories)
    }

    /**
     * Converts the class into a string representation of a json object 
     * 
     * @returns string representation of the object
     */
    toJSONString(): string {
        return JSON.stringify(this)
    }

}

File: Backend/Data/ViewModels/Request/RegenerateRouteVM.cs
namespace Backend.Data.ViewModels.Request
{
    /// <summary>
    /// View Model class for regenerate route reqeust
    /// </summary>
    public class RegenerateRouteVM : GenerateRouteVM
    {
        public List<String> DeletedPlaces { get; set; }

        /// <summary>
        /// Class' constructor method
        /// </summary>
        /// <param name="name">Route's name</param>
        /// <param name="startTime">Time that route starts</param>
        /// <param name="finishTime">Time that route ends</param>
        /// <param name="visitedPlaces">Whether or not the user wants to visit places already visited</param>
        /// <param name="categories">Categories of palces that user chose</param>
        /// <param name="deletedPlaces">List with the places that the user deleted, 
        /// this list is use to know what places the user didn't want in the previous route, so we regenerate different places</param>
        public RegenerateRouteVM(string name, DateTime startTime, DateTime finishTime,
        Boolean visitedPlaces, List<String> categories, List<String> deletedPlaces) : base (name, startTime, finishTime, visitedPlaces, categories)
        {
            this.DeletedPlaces = deletedPlaces;
        }
    }
}